src/compiler/tracing.ts

```diff
@@ -88,38 +93,39 @@ namespace ts.tracing {
         Emit = "emit",
     }
 
-    export function begin(phase: Phase, name: string, args: object) {
-        if (!traceFd) {
-            return;
-        }
-        Debug.assert(fs);
+    export function begin(phase: Phase, name: string, args?: object) {
+        writeEvent("B", phase, name, args);
+    }
 
-        performance.mark("beginTracing");
-        fs.writeSync(traceFd, `{"pid":1,"tid":1,"ph":"B","cat":"${phase}","ts":${1000 * timestamp()},"name":"${name}","args":{ "ts": ${JSON.stringify(args)} }},\n`);
-        performance.mark("endTracing");
-        performance.measure("Tracing", "beginTracing", "endTracing");
+    export function end(phase: Phase, name: string, args?: object) {
+        writeEvent("E", phase, name, args);
     }
 
-    export function end() {
-        if (!traceFd) {
-            return;
-        }
-        Debug.assert(fs);
+    export function instant(phase: Phase, name: string, args?: object) {
+        writeEvent("I", phase, name, args, `"s":"g"`);
+    }
 
-        performance.mark("beginTracing");
-        fs.writeSync(traceFd, `{"pid":1,"tid":1,"ph":"E","ts":${1000 * timestamp()}},\n`);
-        performance.mark("endTracing");
-        performance.measure("Tracing", "beginTracing", "endTracing");
+    // Used for "Complete" (ph:"X") events
+    const completeEvents: { phase: Phase, name: string, args?: object, time: number }[] = [];
+    export function push(phase: Phase, name: string, args?: object) {
+        completeEvents.push({ phase, name, args, time: 1000 * timestamp() });
+    }
+    export function pop() {
+        Debug.assert(completeEvents.length > 0);
+        const { phase, name, args, time } = completeEvents.pop()!;
+        const dur = 1000 * timestamp() - time;
+        writeEvent("X", phase, name, args, `"dur":${dur}`, time);
     }
 
-    export function instant(phase: Phase, name: string, args: object) {
-        if (!traceFd) {
-            return;
-        }
+    function writeEvent(eventType: string, phase: Phase, name: string, args: object | undefined, extras?: string,
+                       time: number = 1000 * timestamp()) {
+        if (!traceFd) return;
         Debug.assert(fs);
-
         performance.mark("beginTracing");
-        fs.writeSync(traceFd, `{"pid":1,"tid":1,"ph":"i","cat":"${phase}","ts":${1000 * timestamp()},"name":"${name}","s":"g","args":{ "ts": ${JSON.stringify(args)} }},\n`);
+        fs.writeSync(traceFd, `,{"pid":1,"tid":1,"ph":"${eventType}","cat":"${phase}","ts":${time},"name":"${name}"`);

```

Comment:
Why did this comma need to move?  If it has to be this way, maybe we could move the `\n` too so that the comma is still at EOL?

Reply:
The resulting dump will look like:

    ,{...}
    ,{...}
    ,{...}
    ,{...}
    ]

so this makes it possible to have decent looking output and not worry about aborting or special casing the last value.  I normally dislike this style too, but I want to avoid special cases to avoid as much runtime overhead as possible, and avoid complicating the code for something that is rarely used (looking at the dump text directly).

Reply:
Right, but what if it was `,\n{"pid"`?  Then the comma still wouldn't be inserted without a following item, but it would appear on the right line.